### PR TITLE
Format dsl

### PR DESCRIPTION
**Pull request by piotrolszewski-a**
## What:

*Format dsl

## Why:

*

## Example:

*

### CODE REVIEW

File: src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * @param args arguments to be inserted into formatted string
+     * @return format operator, evaluated client side.
+     * Operator returns this string with applied format params.
+     * @see: FormatOperationTest
+     */
+    @ClientLogicMarker
+    fun format(formatString: String, vararg args: ClientLogicElement) = FormatOperatorFactory().create(StringElement(formatString), args)
+
+    @ClientLogicMarker
+    fun ClientLogicElement.format(vararg args: ClientLogicElement) = FormatOperatorFactory().create(this, args)
```
Comment (KTGypsy) [inline src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt]: not sure if its okay to only allow ClientLogicElements here. User is not able to use any strings numbers etc, only registryKey or lists which can be initiated explicitly

Comment (KTGypsy) [reply src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt]: it should probably by Any type

File: src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        .build()
+
+    private fun Array<out ClientLogicElement>.toElements(): ListOfClientElements<ClientLogicElement> {
+        val elementsListBuilder = ListOfClientElements.Builder()
+        forEach {
+            when (it) {
+                is StringElement -> elementsListBuilder.add(it)
+                is NumberElement -> elementsListBuilder.add(it)
+                is BooleanElement -> elementsListBuilder.add(it)
+                is ClientLogicElement -> elementsListBuilder.add(it)
```
Comment (KTGypsy) [inline src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt:44]: Check for instance is always 'true'

File: src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    ) = ClientLogicOperator.Builder("format")
+        .add(formatString)
+        .add(args.toElements())
+        .build()
+
+    private fun Array<out ClientLogicElement>.toElements(): ListOfClientElements<ClientLogicElement> {
+        val elementsListBuilder = ListOfClientElements.Builder()
+        forEach {
+            when (it) {
+                is StringElement -> elementsListBuilder.add(it)
```
Comment (KTGypsy) [inline src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt]: these checks should be about primitive types not the CSL ones I think 🤔

**Review by KTGypsy — CHANGES_REQUESTED** [review]

File: src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        args: Array<out Any>
+    ) = ClientLogicOperator.Builder("format")
+        .add(formatString)
+        .add(args.toElements())
+        .build()
+
+    private fun Array<out Any>.toElements(): ListOfClientElements<ClientLogicElement> {
+        val elementsListBuilder = ListOfClientElements.Builder()
+        forEach {
+            when (it) {
```
Comment (PaulinaSadowskaAllegro) [inline src/main/kotlin/pl/allegro/mobile/logic/operators/string/FormatOperation.kt:40]: `when` can be simplified:
```
            when (it) {
                is String -> elementsListBuilder.add(StringElement(it))
                is Number -> elementsListBuilder.add(NumberElement(it))
                is Boolean -> elementsListBuilder.add(BooleanElement(it))
                is ClientLogicElement -> elementsListBuilder.add(it)
            }
```

**Review by PaulinaSadowskaAllegro — CHANGES_REQUESTED** [review]

**Review by KTGypsy — DISMISSED** [review]

**Review by PaulinaSadowskaAllegro — APPROVED** [review]

**Review by KTGypsy — APPROVED** [review]

**Pull request merged**